keel-orca/src/test/kotlin/com/netflix/spinnaker/keel/orca/OrcaTaskMonitorAgentTests.kt

```diff
@@ -1,50 +1,269 @@
     fixture {
       OrcaTaskMonitorAgentFixture(
-        event = TaskCreatedEvent(
-          TaskRecord(
-            id = "123",
-            subject = "titus:cluster:titustestvpc:app-env13",
-            name = "upsert server group")
-        )
+        event = TaskCreatedEvent(taskResourceRecord)
       )
     }
 
-    context("a new task is being created in orca") {
+    context("a new orca task is being stored") {
       before {
-          listener.onTaskEvent(event)
+        subject.onTaskEvent(event)
+        expectThat(repository.getTasks().size).isEqualTo(1)
+        resourceRepository.store(resource)
       }
 
-      test("a new record is being added to the table exactly once") {
-        verify(exactly = 1) { repository.store(any()) }
+      after {
+        clearAllMocks()
+        resourceRepository.dropAll()
+      }
+
+      test("do not process running orca tasks") {
+        coEvery {
+          orcaService.getOrchestrationExecution(event.taskRecord.id)
+        } returns executionDetailResponse(event.taskRecord.id)
+
+        runBlocking {
+          subject.invokeAgent()
+        }
+
+        verify(exactly = 0) { publisher.publishEvent(any()) }
+        expectThat(repository.getTasks().size).isEqualTo(1)
+      }
+
+      test("a task ended with succeeded status") {
+        coEvery {
+          orcaService.getOrchestrationExecution(event.taskRecord.id)
+        } returns executionDetailResponse(event.taskRecord.id, OrcaExecutionStatus.SUCCEEDED)
+
+        runBlocking {
+          subject.invokeAgent()
+        }
+
+        verify(exactly = 1) { publisher.publishEvent(ofType<ResourceTaskSucceeded>()) }
+        expectThat(repository.getTasks().size).isEqualTo(0)
+      }
+
+      test("a task is ended with a failure status with no error") {
+        coEvery {
+          orcaService.getOrchestrationExecution(event.taskRecord.id)
+        } returns executionDetailResponse(event.taskRecord.id, OrcaExecutionStatus.TERMINAL)
+
+        runBlocking {
+          subject.invokeAgent()
+        }
+
+        verify(exactly = 1) { publisher.publishEvent(ofType<ResourceTaskFailed>()) }
+        expectThat(repository.getTasks().size).isEqualTo(0)
+      }
+
+      test("a task is ended with a failure status with a general exception") {
+        coEvery {
+          orcaService.getOrchestrationExecution(event.taskRecord.id)
+        } returns executionDetailResponse(
+          event.taskRecord.id,
+          OrcaExecutionStatus.TERMINAL,
+          OrcaExecutionStages(listOf(orcaContext(exception = orcaExceptions())))
+        )
+
+        runBlocking {
+          subject.invokeAgent()
+        }
+
+        verify(exactly = 1) { publisher.publishEvent(ofType<ResourceTaskFailed>()) }
+        resourceRepository.appendHistory(ResourceTaskFailed(resource, orcaExceptions().details.errors.joinToString(","), clock))
+
+        expectThat(resourceRepository.eventHistory(resource.id))
```

Comment:
This is not actually testing anything because you're explicitly storing that record in the statement immediately above... I noticed that there doesn't seem to be an event listener for `ResourceTaskFailed` like there is for `TaskCreatedEvent` here: https://github.com/spinnaker/keel/pull/777/files#diff-9d2c1a54dd7b764b165eb0aa0ea1d408R37-R40

Wouldn't the expected behavior be to automatically store an event in the history for the failed task?